src/Altinn.App.Core/Helpers/Serialization/ModelSerializationService.cs

```diff
@@ -0,0 +1,265 @@
+    {
+        var modelType = model.GetType();
+        using var span = _telemetry?.StartSerializeToJsonActivity(modelType);
+        var json = JsonSerializer.SerializeToUtf8Bytes(model, modelType, _jsonSerializerOptions);
+        return json;
+    }
+
+    /// <summary>
+    /// Deserialize a single object from a stream
+    /// </summary>
+    public async Task<ServiceResult<object, ProblemDetails>> DeserializeSingleFromStream(
+        Stream body,
+        string? contentType,
+        DataType dataType
+    )
+    {
+        using var memoryStream = new MemoryStream();
+        await body.CopyToAsync(memoryStream);
+        if (!memoryStream.TryGetBuffer(out var segment))
+        {
+            throw new InvalidOperationException("Failed to get buffer from memory stream");
+        }
+
+        var modelType = GetModelTypeForDataType(dataType);
+        object model;
+        if (contentType?.Contains("application/xml") ?? true) // default to xml if no content type is provided
+        {
+            model = DeserializeXml(segment, modelType);
+        }
+        else if (contentType.Contains("application/json"))
+        {
+            model = DeserializeJson(segment, modelType);
+        }
+        else
+        {
+            return new ProblemDetails()
+            {
+                Title = "Unsupported content type",
+                Detail = $"Content type {contentType} is not supported for deserialization",
+                Status = StatusCodes.Status415UnsupportedMediaType,
+            };
+        }
+
+        return model;
+    }
+
+    /// <summary>
+    /// Deserialize utf8 encoded json data to a model of the specified type
+    ///
+    /// Basically just JsonSerializer.Deserialize, but with a telemetry activity and BOM removal
+    /// </summary>
+    /// <param name="data">The binary UTF8 encoded json</param>
+    /// <param name="modelType">The target type for deserialization</param>
+    /// <returns>The deserialized object</returns>
+    public object DeserializeJson(ReadOnlySpan<byte> data, Type modelType)
+    {
+        using var activity = _telemetry?.StartDeserializeFromJsonActivity(modelType);
+        return JsonSerializer.Deserialize(data.RemoveBom(), modelType, _jsonSerializerOptions)
+            ?? throw new JsonException("Json deserialization returned null");
+    }
+
+    /// <summary>
+    /// Deserialize utf8 encoded xml data to a model of the specified type
+    /// </summary>
+    public object DeserializeXml(ReadOnlySpan<byte> data, Type modelType)
+    {
+        using var activity = _telemetry?.StartDeserializeFromXmlActivity(modelType);
+        // convert to UTF16 string as it seems to be preferred by the XmlTextReader
+        // and aligns with previous implementation
+        string streamContent = Encoding.UTF8.GetString(data.RemoveBom());
+        if (string.IsNullOrWhiteSpace(streamContent))
+        {
+            throw new Exception("No XML content read from stream");
```

Comment:
Bruke ArgumentException?

Reply:
Vet ikke helt om jeg ser poenget gitt at vi er enige om å ikke styre controlFlow med exceptions.
```suggestion
            throw new ArgumentException("No XML content read from stream");
```